File: packages/python/onnx9000-core/src/onnx9000/core/models/llama.py
"""Llama."""

from typing import Any, Optional

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, mul
from onnx9000.core.primitives import (
    Gemm,
    GroupedQueryAttention,
    RMSNorm,
    RoPE,
    Silu,
)


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    """Get param."""
    return Variable(name=name, shape=shape, dtype=dtype)


class SwiGLU:  # noqa: D101
    """Swi glu."""

    def __init__(self, hidden_dim: int, ffn_dim: int, prefix: str = ""):  # noqa: D107
        """Init."""
        self.prefix = prefix
        self.hidden_dim = hidden_dim
        self.ffn_dim = ffn_dim
        self.w1 = Gemm(trans_b=1)  # Gate
        self.w2 = Gemm(trans_b=1)  # Down
        self.w3 = Gemm(trans_b=1)  # Up
        self.act = Silu()

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        gate = self.w1(x, get_param(f"{self.prefix}.w1.weight", [self.ffn_dim, self.hidden_dim]))
        up = self.w3(x, get_param(f"{self.prefix}.w3.weight", [self.ffn_dim, self.hidden_dim]))

        # SwiGLU = Silu(x * w1) * (x * w3)
        activated_gate = self.act(gate)
        hidden = mul(activated_gate, up)

        down = self.w2(
            hidden, get_param(f"{self.prefix}.w2.weight", [self.hidden_dim, self.ffn_dim])
        )
        return down


class LLaMABlock:  # noqa: D101
    """L la ma block."""

    def __init__(self, dim: int, num_heads: int, num_kv_heads: int, ffn_dim: int, prefix: str = ""):  # noqa: D107
        """Init."""
        self.prefix = prefix
        self.dim = dim
        self.norm1 = RMSNorm((dim,))
        self.attn = GroupedQueryAttention(num_heads, num_kv_heads, qkv_bias=False, out_bias=False)
        self.norm2 = RMSNorm((dim,))
        self.mlp = SwiGLU(dim, ffn_dim, prefix=f"{prefix}.mlp")

    def __call__(self, x: Tensor, pos: Tensor, mask: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        identity = x
        x_norm = self.norm1(x, get_param(f"{self.prefix}.norm1.weight", [self.dim]))

        # In real LLaMA, RoPE is applied to Q and K inside attention, but for AST layout
        # we might just use the primitive. Let's assume GroupedQueryAttention handles it or we apply it before.
        # We will just pass it to attn.
        x_attn = self.attn(x_norm, x_norm, x_norm, mask=mask)
        x = add(identity, x_attn)

        identity = x
        x_norm = self.norm2(x, get_param(f"{self.prefix}.norm2.weight", [self.dim]))
        x_mlp = self.mlp(x_norm)
        x = add(identity, x_mlp)

        return x


class LLaMA:  # noqa: D101
    """L la ma."""

    def __init__(  # noqa: D107
        self,
        vocab_size: int = 32000,
        dim: int = 4096,
        num_heads: int = 32,
        num_kv_heads: int = 32,
        depth: int = 32,
        ffn_dim: int = 11008,
        max_seq_len: int = 2048,
    ):
        """Init."""
        self.vocab_size = vocab_size
        self.dim = dim
        self.depth = depth
        self.max_seq_len = max_seq_len

        # We use a Gather for embedding
        self.blocks = [
            LLaMABlock(dim, num_heads, num_kv_heads, ffn_dim, prefix=f"blocks.{i}")
            for i in range(depth)
        ]
        self.norm = RMSNorm((dim,))
        self.lm_head = Gemm(trans_b=1)
        self.rope = RoPE(dim // num_heads, max_seq_len=max_seq_len)

    def __call__(self, input_ids: Tensor, pos: Tensor, mask: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import gather

        x = gather(
            get_param("tok_embeddings.weight", [self.vocab_size, self.dim]), input_ids, axis=0
        )

        # Apply RoPE to embeddings conceptually or inside blocks
        x = self.rope(x, pos)

        for block in self.blocks:
            x = block(x, pos, mask)

        x = self.norm(x, get_param("norm.weight", [self.dim]))
        x = self.lm_head(x, get_param("output.weight", [self.vocab_size, self.dim]))
        return x


def llama_7b(**kwargs: Any) -> LLaMA:  # noqa: D103
    """Llama 7b."""
    return LLaMA(
        vocab_size=32000, dim=4096, num_heads=32, num_kv_heads=32, depth=32, ffn_dim=11008, **kwargs
    )


def mistral_7b(**kwargs: Any) -> LLaMA:  # noqa: D103
    """Mistral 7b."""
    return LLaMA(
        vocab_size=32000, dim=4096, num_heads=32, num_kv_heads=8, depth=32, ffn_dim=14336, **kwargs
    )

File: packages/js/core/src/models/llama.ts
/* eslint-disable */
import { Tensor } from '../ir/tensor.js';
import { Gemm, GroupedQueryAttention, RMSNorm, RoPE, Silu } from '../primitives.js';

function getParam(
  name: string,
  shape: number[],
  dtype: ReturnType<typeof JSON.parse> = 'float32',
): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: ReturnType<typeof JSON.parse>): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class SwiGLU {
  public prefix: string;
  public hiddenDim: number;
  public ffnDim: number;
  public w1: Gemm;
  public w2: Gemm;
  public w3: Gemm;
  public act: Silu;

  constructor(hiddenDim: number, ffnDim: number, prefix: string = '') {
    this.prefix = prefix;
    this.hiddenDim = hiddenDim;
    this.ffnDim = ffnDim;
    this.w1 = new Gemm(1.0, 1.0, 0, 1);
    this.w2 = new Gemm(1.0, 1.0, 0, 1);
    this.w3 = new Gemm(1.0, 1.0, 0, 1);
    this.act = new Silu();
  }

  call(x: Tensor): Tensor {
    const gate = this.w1.call(
      x,
      getParam(`${this.prefix}.w1.weight`, [this.ffnDim, this.hiddenDim]),
    );
    const up = this.w3.call(x, getParam(`${this.prefix}.w3.weight`, [this.ffnDim, this.hiddenDim]));

    const activatedGate = this.act.call(gate);
    const hidden = recordOp('Mul', [activatedGate, up]);

    const down = this.w2.call(
      hidden,
      getParam(`${this.prefix}.w2.weight`, [this.hiddenDim, this.ffnDim]),
    );
    return down;
  }
}

export class LLaMABlock {
  public prefix: string;
  public dim: number;
  public norm1: RMSNorm;
  public attn: GroupedQueryAttention;
  public norm2: RMSNorm;
  public mlp: SwiGLU;

  constructor(
    dim: number,
    numHeads: number,
    numKvHeads: number,
    ffnDim: number,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.dim = dim;
    this.norm1 = new RMSNorm([dim]);
    this.attn = new GroupedQueryAttention(numHeads, numKvHeads, false, false);
    this.norm2 = new RMSNorm([dim]);
    this.mlp = new SwiGLU(dim, ffnDim, `${prefix}.mlp`);
  }

  call(x: Tensor, pos: Tensor, mask?: Tensor): Tensor {
    let identity = x;
    let xNorm = this.norm1.call(x, getParam(`${this.prefix}.norm1.weight`, [this.dim]));
    const xAttn = this.attn.call(xNorm, xNorm, xNorm, mask);
    x = recordOp('Add', [identity, xAttn]);

    identity = x;
    xNorm = this.norm2.call(x, getParam(`${this.prefix}.norm2.weight`, [this.dim]));
    const xMlp = this.mlp.call(xNorm);
    x = recordOp('Add', [identity, xMlp]);

    return x;
  }
}

export class LLaMA {
  public vocabSize: number;
  public dim: number;
  public depth: number;
  public maxSeqLen: number;
  public blocks: LLaMABlock[];
  public norm: RMSNorm;
  public lmHead: Gemm;
  public rope: RoPE;

  constructor(
    vocabSize: number = 32000,
    dim: number = 4096,
    numHeads: number = 32,
    numKvHeads: number = 32,
    depth: number = 32,
    ffnDim: number = 11008,
    maxSeqLen: number = 2048,
  ) {
    this.vocabSize = vocabSize;
    this.dim = dim;
    this.depth = depth;
    this.maxSeqLen = maxSeqLen;

    this.blocks = [];
    for (let i = 0; i < depth; i++) {
      this.blocks.push(new LLaMABlock(dim, numHeads, numKvHeads, ffnDim, `blocks.${i}`));
    }
    this.norm = new RMSNorm([dim]);
    this.lmHead = new Gemm(1.0, 1.0, 0, 1);
    this.rope = new RoPE(Math.floor(dim / numHeads), 10000.0, maxSeqLen);
  }

  call(inputIds: Tensor, pos: Tensor, mask?: Tensor): Tensor {
    let x = recordOp(
      'Gather',
      [getParam('tok_embeddings.weight', [this.vocabSize, this.dim]), inputIds],
      { axis: 0 },
    );
    x = this.rope.call(x, pos);

    for (const block of this.blocks) {
      x = block.call(x, pos, mask);
    }

    x = this.norm.call(x, getParam('norm.weight', [this.dim]));
    x = this.lmHead.call(x, getParam('output.weight', [this.vocabSize, this.dim]));
    return x;
  }
}

export function llama7b(): LLaMA {
  return new LLaMA(32000, 4096, 32, 32, 32, 11008);
}
